File: backend/project-service/src/main/java/com/gl/projectService/service/ProjectService.java
package com.gl.projectService.service;

import com.gl.projectService.dto.ProjectRequestDTO;
import com.gl.projectService.dto.ProjectResponseDTO;
import com.gl.projectService.dto.StoryRequestDTO;
import com.gl.projectService.exception.ProjectNotFoundException;

import java.util.List;

public interface ProjectService  {
    // US-005
    ProjectResponseDTO createProject(Long founderId, ProjectRequestDTO request);
    ProjectResponseDTO createStory(Long founderId, StoryRequestDTO request);

    // US-006
    List<ProjectResponseDTO> getStoryFeed();
    List<ProjectResponseDTO> getProjectOpenings();

    // US-007
    ProjectResponseDTO getProjectById(Long id) throws ProjectNotFoundException;

    // US-008
    void deleteProject(Long id, Long requestingFounderId) throws ProjectNotFoundException;
}

File: frontend/src/api/projectService.ts
import axiosClient from "./axiosClient";

export enum PostType {
  PROJECT = "PROJECT",
  STORY = "STORY",
}

export interface ProjectResponseDTO {
  id: number;
  founderId: number;
  title: string;
  description: string;
  technologyStack: string;
  postType: PostType;
  totalOpenings: number;
  filledOpenings: number;
  isActive: boolean;
  createdAt: string; // LocalDateTime comes as an ISO string in JSON
  updatedAt: string;
}

export const projectService = {
  // POST /api/projects
  // The founderId is injected by the Gateway
  createProject: async (projectData: any): Promise<ProjectResponseDTO> => {
    const response = await axiosClient.post("/api/projects", projectData);
    return response.data;
  },

  // POST /api/projects/stories
  createStory: async (storyData: any): Promise<ProjectResponseDTO> => {
    const response = await axiosClient.post("/api/projects/stories", storyData);
    return response.data;
  },

  // GET /api/projects/feed
  getStoryFeed: async (): Promise<ProjectResponseDTO[]> => {
    const response = await axiosClient.get("/api/projects/feed");
    return response.data;
  },

  // GET /api/projects/openings
  getProjectOpenings: async (): Promise<ProjectResponseDTO[]> => {
    const response = await axiosClient.get("/api/projects/openings");
    return response.data;
  },

  // GET /api/projects/{id}
  getProjectById: async (id: number): Promise<ProjectResponseDTO> => {
    const response = await axiosClient.get(`/api/projects/${id}`);
    return response.data;
  },

  // DELETE /api/projects/{id}
  // The Gateway ensures only the owner (matching founderId) can delete
  deleteProject: async (id: number): Promise<string> => {
    const response = await axiosClient.delete(`/api/projects/${id}`);
    return response.data;
  },
};
